timApp/tide_support/utils.py

```diff
@@ -0,0 +1,504 @@
+    )
+
+    user_ctx = UserContext.from_one_user(u=user)
+
+    pars = doc.document.get_paragraphs()
+
+    tasks = []
+
+    for p in pars:
+        if p.attrs.get(IDE_TASK_TAG) is not None:
+            task = ide_user_plugin_data(
+                doc=doc, par=p, user_ctx=user_ctx, ide_task_id=p.attrs.get(IDE_TASK_TAG)
+            )
+            if task:
+                tasks.append(task)
+
+    if len(tasks) == 0:
+        return TIDEError("No tasks found")
+
+    return tasks
+
+
+def ide_task_by_id(
+        user: User,
+        ide_task_id: str,
+        doc_id: int = None,
+        doc_path: str = None,
+) -> TIDEPluginData | TIDEError:
+    """
+    Get TIDE-task from the document by document id and paragraph id
+    :param ide_task_id:  TIDE-task id
+    :param user: Logged-in user
+    :param doc_id:  Document id
+    :param doc_path:  Document path
+    :return: JSON with TIDE-task ide-files, task info and task id
+    """
+    if doc_id is None and doc_path is None:
+        return TIDEError(error="No document id or path given")
+
+    if doc_path is None:
+        doc = DocInfo.find_by_id(item_id=doc_id)
+    else:
+        doc = DocInfo.find_by_path(path=doc_path.lower())
+
+    # If the document does not exist, raise NotExist
+    if doc is None:
+        return TIDEError("No document found")
+
+    # Check if the user has edit access to the document
+
+    verify_access(
+        b=doc,
+        access_type=AccessType(2),
+        require=True,
+        message="You do not have access to this document",
+        user=user,
+    )
+
+    user_ctx = UserContext.from_one_user(u=user)
+
+    pars = doc.document.get_paragraphs()
+
+    if pars is None:
+        return TIDEError("No paragraphs found")
+
+    tasks = []
+
+    for p in pars:
+        if p.attrs.get(IDE_TASK_TAG) == ide_task_id:
+            task = ide_user_plugin_data(
+                doc=doc, par=p, user_ctx=user_ctx, ide_task_id=ide_task_id
+            )
+            if task:
+                tasks.append(task)
+
+    if len(tasks) == 0:
+        return TIDEError("No tasks found")
+
+    if len(tasks) == 1:
+        return tasks[0]
+
+    # TODO: case where files are saved based on language base folders eg path is taken from the language package
+
+    return TIDEError("Multiple tasks found, support not implemented yet")
+
+
+def ide_user_plugin_data(
+        doc: DocInfo,
+        par: dict,
+        ide_task_id: str,
+        user_ctx: UserContext,
+) -> TIDEPluginData | TIDEError:
+    """
+    Get the TIDE-task information from the plugin
+    :param ide_task_id:  TIDE-task id
+    :param doc: TIM document
+    :param par: Paragraph from the document
+    :param user_ctx: User context
+    :return: JSON with TIDE-task ide-files, task info and task id
+    """
+
+    view_ctx = default_view_ctx
+
+    plugin = Plugin.from_paragraph(par, view_ctx, user_ctx)
+
+    if plugin.type != "csPlugin":
+        return TIDEError(error="Not a csPlugin plugin")
+
```

Comment:
Ehdotan, että virheiden käsittely tehtäisiin Pythonin tapaan poikkeuksilla, jotta tuota palautustyyppiä ei tarvitse määrittää vaikeasti. Eli siis TIDEError voisi olla ihan poikkeus (eli periä esim. `Exception`-luokkaa), jolloin sitä voisi käsitellä muodossa

```suggestion
        raise TIDEError(error="Not a csPlugin plugin")
```

TAI TIDEError voi periä `RouteException`ia, jolloin Flask automaattisesti palauttaa virheen JSONina.

Reply:
Muokattu perimään `RouteException`:n.